### PR TITLE
Init user settings page

### PR DESCRIPTION
**Pull request by zacharyblasczyk**

### CODE REVIEW

File: apps/webservice/src/app/[workspaceSlug]/SidebarWorkspaceDropdown.tsx [hunk 1]
```diff
@@ -27,7 +27,7 @@ export const SidebarWorkspaceDropdown: React.FC<{ workspace: Workspace }> = ({
 }) => {
   const { data } = useSession();
   const workspaces = api.workspace.list.useQuery();
-  const update = api.profile.update.useMutation();
+  const update = api.profile.updateActiveWorkspace.useMutation();
```
Comment (jsbroks) [inline apps/webservice/src/app/[workspaceSlug]/SidebarWorkspaceDropdown.tsx:30]: why? Don't you need to pass in a workspaceId?

Comment (zacharyblasczyk) [reply apps/webservice/src/app/[workspaceSlug]/SidebarWorkspaceDropdown.tsx:30]: Updating the username has nothing to do with the workspace. I can change it back if you want.

Comment (zacharyblasczyk) [reply apps/webservice/src/app/[workspaceSlug]/SidebarWorkspaceDropdown.tsx:30]: This was a side effect of me changing the api endpoint.

File: packages/api/src/router/user.ts [hunk 2]
```diff
@@ -20,6 +20,16 @@ export const profileRouter = createTRPCRouter({
...
+  updateUsername: protectedProcedure
+    .input(z.object({ username: z.string().min(3).max(30) }))
+    .mutation(({ ctx, input }) =>
+      ctx.db
+        .update(user)
+        .set({ username: input.username })
+        .where(eq(user.id, ctx.session.user.id))
+        .returning()
+        .then(takeFirst),
+    ),
```
Comment (jsbroks) [inline packages/api/src/router/user.ts:32]: this should be under the same update function. it doesn't need to be its own endpoint

File: apps/webservice/src/app/[workspaceSlug]/settings/(settings)/layout.tsx [hunk 3]
```diff
@@ -3,5 +3,5 @@ export default function AccountSettingsLayout({
 }: {
   children: React.ReactNode;
 }) {
-  return <div className="m-10 mt-20">{children}</div>;
+  return <div className="mt-10">{children}</div>;
```
Comment (zacharyblasczyk) [inline apps/webservice/src/app/[workspaceSlug]/settings/(settings)/layout.tsx:6]: This was a weird margin to have. It looks better this way on all the user setting pages.

File: apps/webservice/src/app/[workspaceSlug]/settings/(settings)/account/profile/page.tsx [hunk 4]
```diff
@@ -1,15 +1,142 @@
...
+            </li>
+          </section>
+          <section className="space-y-6 rounded-b-lg border border-neutral-800/70 bg-background px-6 shadow-lg">
+            {/* Email */}
+            <li className="border-b py-6 lg:mt-0">
+              <div className="flex flex-row items-center justify-between">
+                <Label>Email</Label>
+                <Input readOnly value={email} className="w-64" />
+              </div>
+            </li>
```
Comment (jsbroks) [inline apps/webservice/src/app/[workspaceSlug]/settings/(settings)/account/profile/page.tsx:60]: the label says email??

File: apps/webservice/src/app/[workspaceSlug]/settings/(settings)/account/profile/page.tsx [hunk 5]
```diff
@@ -1,15 +1,142 @@
...
+    <div className="scrollbar-thin scrollbar-thumb-neutral-800 scrollbar-track-neutral-900 h-[calc(100vh-120px)] overflow-auto">
+      <div className="container mx-auto max-w-2xl space-y-8 py-8">
+        <div className="space-y-1">
+          <h1 className="text-xl font-semibold">Profile</h1>
+          <p className="text-sm text-muted-foreground">
+            Manage your Ctrlplane profile
+          </p>
+        </div>
+
+        {/* Profile Section */}
```
Comment (jsbroks) [inline apps/webservice/src/app/[workspaceSlug]/settings/(settings)/account/profile/page.tsx:32]: title says profile section

Comment (jsbroks) [reply apps/webservice/src/app/[workspaceSlug]/settings/(settings)/account/profile/page.tsx:32]: If you need these comments move them into components

File: apps/webservice/src/app/[workspaceSlug]/settings/(settings)/account/profile/page.tsx [hunk 6]
```diff
@@ -1,15 +1,142 @@
...
+import { Input } from "@ctrlplane/ui/input";
+import { Label } from "@ctrlplane/ui/label";
+
+import { api } from "~/trpc/server";
+import UpsertUsername from "./UpsertUsername";
+
+const defaultAvatar = "/apple-touch-icon.png";
+
+export default async function AccountSettingProfilePage() {
+  const session = await auth();
```
Comment (jsbroks) [inline apps/webservice/src/app/[workspaceSlug]/settings/(settings)/account/profile/page.tsx:17]: does session not have user info?

File: apps/webservice/src/app/[workspaceSlug]/settings/(settings)/account/profile/UpsertUsername.tsx [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+import type { User } from "@ctrlplane/db/schema";
+import type { ChangeEvent } from "react";
+import { useState } from "react";
+
+import { Input } from "@ctrlplane/ui/input";
+import { toast } from "@ctrlplane/ui/toast";
+
+import { api } from "~/trpc/react";
+
+export default function UpsertUsername({
```
Comment (jsbroks) [inline apps/webservice/src/app/[workspaceSlug]/settings/(settings)/account/profile/UpsertUsername.tsx:12]: do not export default

File: apps/webservice/src/app/[workspaceSlug]/settings/(settings)/account/api/page.tsx [hunk 8]
```diff
@@ -8,21 +8,25 @@ export const metadata: Metadata = { title: "API Key Creation" };
...
-        <h1 className="text-xl font-semibold">API</h1>
-      </div>
-      <div className="border-b" />
-      <div className="space-y-6">
-        <div>General</div>
-        <div className="text-sm text-muted-foreground">
-          You can create personal API keys for accessing Ctrlplane's API to
-          build your own integration or hacks.
-        </div>
+    <>
```
Comment (jsbroks) [inline apps/webservice/src/app/[workspaceSlug]/settings/(settings)/account/api/page.tsx:11]: why the fragment

File: apps/webservice/src/app/[workspaceSlug]/settings/(settings)/account/profile/page.tsx [hunk 9]
```diff
@@ -1,15 +1,142 @@
...
+                <SiGithub className="mr-2 h-4 w-4" />
+                <span className="w-96">
+                  <Label>GitHub</Label>
+                  <p className="w-full text-sm text-gray-400">
+                    Link Pull Requests with your account.
+                  </p>
+                </span>
+                <Button variant="outline" className="w-32">
+                  Connect
+                </Button>
```
Comment (jsbroks) [inline apps/webservice/src/app/[workspaceSlug]/settings/(settings)/account/profile/page.tsx:119]: reword and connect this to our github

File: apps/webservice/src/app/[workspaceSlug]/settings/(settings)/account/api/page.tsx (hunk 8 above)
Comment (zacharyblasczyk) [reply apps/webservice/src/app/[workspaceSlug]/settings/(settings)/account/api/page.tsx:11]: That was a leftover from my testing.

File: apps/webservice/src/app/[workspaceSlug]/settings/(settings)/account/profile/page.tsx (hunk 4 above)
Comment (zacharyblasczyk) [reply apps/webservice/src/app/[workspaceSlug]/settings/(settings)/account/profile/page.tsx]: Screenshot 2024-09-19 at 11 04 54 PM

Yes it does.

File: apps/webservice/src/app/[workspaceSlug]/settings/(settings)/account/profile/page.tsx (hunk 6 above)
Comment (zacharyblasczyk) [reply apps/webservice/src/app/[workspaceSlug]/settings/(settings)/account/profile/page.tsx:17]: It doesn't, just the id. 
```tsx
declare module "next-auth" {
  interface Session {
    token: JWT;
    user: {
      id: string;
    } & DefaultSession["user"];
  }
}
```

And DefaultSession only has:
```tsx
export interface User {
  id?: string
  name?: string | null
  email?: string | null
  image?: string | null
}
```

File: apps/webservice/src/app/[workspaceSlug]/settings/(settings)/account/profile/UpsertUsername.tsx [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+import { api } from "~/trpc/react";
+
+export const UpsertUsername: React.FC<{
+  user: User;
+  className?: string;
+}> = ({ user, className }) => {
+  const [username, setUsername] = useState(user.username);
+  const [cachedUsername, setCachedUsername] = useState(user.username);
+  const updateProfile = api.profile.update.useMutation();
```
Comment (jsbroks) [inline apps/webservice/src/app/[workspaceSlug]/settings/(settings)/account/profile/UpsertUsername.tsx:18]:
```suggestion
  const update = api.profile.update.useMutation();
```

File: packages/api/src/router/user.ts [hunk 11]
```diff
@@ -10,12 +10,20 @@ import { createTRPCRouter, protectedProcedure } from "../trpc";
 export const profileRouter = createTRPCRouter({
   update: protectedProcedure
     .input(
-      z.object({ activeWorkspaceId: z.string().uuid().nullable().optional() }),
+      z.object({
+        activeWorkspaceId: z.string().uuid().nullable().optional(),
+        username: z.string().min(3).max(30).optional(),
```
Comment (jsbroks) [inline packages/api/src/router/user.ts:15]: can you create this from the zod drizzle integration so we don't need to manually define it

File: packages/db/src/schema/auth.ts [hunk 12]
```diff
@@ -16,6 +16,7 @@ import { workspace } from "./workspace.js";
 export const user = pgTable("user", {
   id: uuid("id").notNull().primaryKey().defaultRandom(),
   name: varchar("name", { length: 255 }),
+  username: varchar("username", { length: 255 }),
```
Comment (jsbroks) [inline packages/db/src/schema/auth.ts:19]: probably need a uniqueness constraint on this?

Comment (jsbroks) [reply packages/db/src/schema/auth.ts:19]: should also be not nullable

Comment (jsbroks) [reply packages/db/src/schema/auth.ts:19]: I think its best we just remove username for now

File: apps/webservice/src/app/[workspaceSlug]/SidebarWorkspaceDropdown.tsx [hunk 13]
```diff
@@ -60,7 +60,12 @@ export const SidebarWorkspaceDropdown: React.FC<{ workspace: Workspace }> = ({
                   key={ws.id}
                   href={`/${ws.slug}`}
                   passHref
-                  onClick={() => update.mutate({ activeWorkspaceId: ws.id })}
+                  onClick={() =>
+                    update.mutate({
+                      email: data?.user.email ?? "",
```
Comment (jsbroks) [inline apps/webservice/src/app/[workspaceSlug]/SidebarWorkspaceDropdown.tsx:65]:
```suggestion
```

we shouldn't need to pass this in

File: apps/webservice/src/app/[workspaceSlug]/settings/(settings)/account/profile/GithubRedirectButton.tsx [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+  className?: string;
+  githubUserId: string;
+}> = ({ variant, className, githubUserId }) => {
+  const router = useRouter();
+
+  return (
+    <Button
+      variant={variant}
+      className={className}
+      onClick={() => router.push("../workspace/integrations/github")}
```
Comment (jsbroks) [inline apps/webservice/src/app/[workspaceSlug]/settings/(settings)/account/profile/GithubRedirectButton.tsx:18]:
```suggestion
      onClick={() => router.push("../workspace/integrations/github")}
```

absolute path

File: packages/api/src/router/user.ts [hunk 15]
```diff
@@ -3,19 +3,17 @@ import { z } from "zod";
...
   update: protectedProcedure
-    .input(
-      z.object({ activeWorkspaceId: z.string().uuid().nullable().optional() }),
-    )
+    .input(createUser)
     .mutation(({ ctx, input }) =>
       ctx.db
         .update(user)
-        .set(input)
+        .set(createUser.parse(input))
```
Comment (jsbroks) [inline packages/api/src/router/user.ts:16]:
```suggestion
        .set(input)
```

trpc already parses this

File: packages/db/src/schema/auth.ts [hunk 16]
```diff
@@ -26,6 +28,12 @@ export const user = pgTable("user", {
 
 export type User = InferSelectModel<typeof user>;
 
+export const createUser = createInsertSchema(user, {
+  name: z.string().min(1).max(255),
+  email: z.string().email(),
+  activeWorkspaceId: z.string().uuid().optional(),
+}).omit({ id: true });
```
Comment (jsbroks) [inline packages/db/src/schema/auth.ts:35]: also add updateUser

Comment (jsbroks) [reply packages/db/src/schema/auth.ts:35]: follow the pattern

**Review by jsbroks — APPROVED** [review]

**Pull request merged**